### ISSUE TITLE
Federation Gateway resolving wrong field during buildQueryPlan

### ISSUE THREAD

**Issue by WardGubbi** [apollographql/apollo-server#3415]
I'm transforming my Apollo GraphQL into a federated one and I'm seeing an issue with property types in my inline fragments being resolved incorrectly and throwing an error on the gateway

The query in question
```
{
  grid(
    brand: "example.com"
    path: "C4054A88-1091-41B9-A53E-0AB2706DB2B8"
    pageNumber: 1
    displaySize: SMALL
  ) {
    cells {
      __typename
      ... on GridSlider {
        spanSize
      }

      ... on GridTeaser {
        spanSize
        config {
          type
          alignment
          ratio
          numbered
          stripped
        }
      }
      
      ... on GridHtml {
        spanSize
        config {
          url
          html
        }
      }

      ... on GridBanner {
        spanSize
        config {
          adUnit
          pageType
          slot
          format
        }
      }

      ... on GridHeader {
        spanSize
        config {
          title
          color
        }
      }
    }
  }
}
```

This results in an error on the gateway
```
"GraphQLError: Cannot query field \"type\" on type \"GridTeaserConfig\"",
"    at QueryPlanningContext.getFieldDef (/Users/Ward/Projecten/mediahuis/federation-repro/node_modules/@apollo/gateway/dist/buildQueryPlan.js:381:19)",
"    at collectFields (/Users/Ward/Projecten/mediahuis/federation-repro/node_modules/@apollo/gateway/dist/buildQueryPlan.js:256:42)",
```

The same request does work when fed to the service itself.

In my query it will try to resolve `config.url` from `GridHtml` as a property of `GridTeaserConfig` (note that the previous fragment was of type `GridTeaser`) instead of treating config as `GridHtmlConfig`. When removing the `GridTeaser` fragment (or removing its config) the issue repeats itself on the next config, `config.adUnit` being resolved as property of `GridHtmlConfig` instead of `GridBannerConfig`. Reordering (or deleting some fragments) shows that it always seems to resolve to the previous type's config instead of the correct one. Keeping only one of the configs (the HTML one for example) makes it work as expected.

I'm seeing this in @apollo/gateway 0.10.4

Here is a repository (adjusted demo repo) that reproduces the behavior: https://github.com/WardGubbi/graphql-federation

**Comment by thicit** [comment]
This issue seems to originate in the groups stage of the build query plan.
The field collection beforehand resolves correctly, however in the groups stage things go wrong.
completeFields & collectSubfields take the first field without considering parentType.

**Comment by geoffreysamper** [comment]
any news arround this issue

**Comment by geoffreysamper** [comment]
@thicit do you have any idea what must be done to fix this issue and get it merged?

**Comment by geoffreysamper** [comment]
This issue is fixed in https://github.com/apollographql/apollo-server/pull/3581.

**Comment by abernix** [comment]
Fixed by #3581.  Please report back if this isn't the case!
